File: buzu/src/main/java/br/net/buzu/context/BasicSubtypeManager.java
/*
 *	This file is part of Buzu.
 *
 *   Buzu is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU Lesser General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   Buzu is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU Lesser General Public License for more details.
 *
 *   You should have received a copy of the GNU Lesser General Public License
 *   along with Buzu.  If not, see <http://www.gnu.org/licenses/>.
 */
package br.net.buzu.context;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.time.Period;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;

import br.net.buzu.pplspec.context.SubtypeManager;
import br.net.buzu.pplspec.model.Subtype;

/**
 * Stateless SubType Manager.
 *
 * @author Douglas Siviotti
 * @since 1.0
 *
 */
public class BasicSubtypeManager implements SubtypeManager {

	public static final SubtypeManager INSTANCE = new BasicSubtypeManager();

	static final String SUBTYPE_NOT_FOUND = "Subtype not found:";
	private static final Map<Class<?>, Subtype> INTERNAL_MAP = new HashMap<>();
	static {
		// Simple Types
		INTERNAL_MAP.put(String.class, Subtype.STRING);
		INTERNAL_MAP.put(Integer.class, Subtype.INTEGER);
		INTERNAL_MAP.put(BigInteger.class, Subtype.INTEGER);
		INTERNAL_MAP.put(int.class, Subtype.INTEGER);
		INTERNAL_MAP.put(Boolean.class, Subtype.BOOLEAN);
		INTERNAL_MAP.put(boolean.class, Subtype.BOOLEAN);
		INTERNAL_MAP.put(Date.class, Subtype.TIMESTAMP);
		INTERNAL_MAP.put(LocalDateTime.class, Subtype.TIMESTAMP);
		INTERNAL_MAP.put(LocalDate.class, Subtype.DATE);
		INTERNAL_MAP.put(LocalTime.class, Subtype.TIME);
		INTERNAL_MAP.put(Period.class, Subtype.PERIOD);
		INTERNAL_MAP.put(BigDecimal.class, Subtype.NUMBER);
		INTERNAL_MAP.put(Double.class, Subtype.NUMBER);
		INTERNAL_MAP.put(double.class, Subtype.NUMBER);
		INTERNAL_MAP.put(Float.class, Subtype.NUMBER);
		INTERNAL_MAP.put(float.class, Subtype.NUMBER);
		INTERNAL_MAP.put(Long.class, Subtype.LONG);
		INTERNAL_MAP.put(BigInteger.class, Subtype.LONG);
		INTERNAL_MAP.put(long.class, Subtype.LONG);
		INTERNAL_MAP.put(Byte.class, Subtype.INTEGER);
		INTERNAL_MAP.put(byte.class, Subtype.INTEGER);
		INTERNAL_MAP.put(Short.class, Subtype.INTEGER);
		INTERNAL_MAP.put(short.class, Subtype.INTEGER);
		INTERNAL_MAP.put(Character.class, Subtype.CHAR);
		INTERNAL_MAP.put(char.class, Subtype.CHAR);
	}

	/**
	 * Creates a instance of <code>SubType</code> based on a Class.
	 * 
	 * @param typeClass
	 *            The class
	 * @return The correct instance relative to the class.
	 */
	public Subtype fromType(Class<?> typeClass) {
		if (typeClass.isEnum()) {
			return Subtype.STRING;
		}
		Subtype simpleType = INTERNAL_MAP.get(typeClass);
		return simpleType != null ? simpleType : Subtype.OBJ;
	}

	public boolean isSimple(Class<?> type) {
		return INTERNAL_MAP.containsKey(type) || type.isEnum();
	}

	/**
	 * Creates a instance of <code>SubType</code> based on a text
	 * representation.
	 * 
	 * @param text
	 *            The text representation
	 * @param indicates
	 *            if the Metadata is complex
	 * @return The correct instance relative to text or <code>null</code> if the
	 *         text is invalid.
	 */
	public Subtype fromText(String text, boolean complex) {
		if (text == null || text.isEmpty()) {
			return complex ? Subtype.DEFAULT_COMPLEX : Subtype.DEFAULT_SINGLE;
		}
		return Subtype.get(text);
	}
}

File: ppl-lib/src/main/kotlin/br/net/buzu/pplimpl/jvm/BasicSubtypeManager.kt
/*
 *	This file is part of Buzu.
 *
 *   Buzu is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU Lesser General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   Buzu is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty domainOf
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU Lesser General Public License for more details.
 *
 *   You should have received a copy of the GNU Lesser General Public License
 *   along with Buzu.  If not, see <http://www.gnu.org/licenses/>.
 */
package br.net.buzu.pplimpl.jvm

import br.net.buzu.pplspec.context.SubtypeManager
import br.net.buzu.pplspec.model.Subtype

import java.math.BigDecimal
import java.math.BigInteger
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.Period
import java.util.Date
import java.util.HashMap

internal val SUBTYPE_NOT_FOUND = "Subtype not found:"

private val DEFAULT_COMPLEX = Subtype.DEFAULT_COMPLEX
private val DEFAULT_SINGLE = Subtype.DEFAULT_SINGLE

private val INTERNAL_MAP :Map<Class<*>, Subtype> = mapOf(
        String::class.java to Subtype.STRING,
        Int::class.java to Subtype.INTEGER,
        BigInteger::class.java to Subtype.INTEGER,
        Int::class.javaPrimitiveType!! to Subtype.INTEGER,
        Boolean::class.java to Subtype.BOOLEAN,
        Boolean::class.javaPrimitiveType!! to Subtype.BOOLEAN,
        Date::class.java to Subtype.TIMESTAMP,
        LocalDateTime::class.java to Subtype.TIMESTAMP,
        LocalDate::class.java to Subtype.DATE,
        LocalTime::class.java to Subtype.TIME,
        Period::class.java to Subtype.PERIOD,
        BigDecimal::class.java to Subtype.NUMBER,
        Double::class.java to Subtype.NUMBER,
        Double::class.javaPrimitiveType!! to Subtype.NUMBER,
        Float::class.java to Subtype.NUMBER,
        Float::class.javaPrimitiveType!! to Subtype.NUMBER,
        Long::class.java to Subtype.LONG,
        BigInteger::class.java to Subtype.LONG,
        Long::class.javaPrimitiveType!! to Subtype.LONG,
        Byte::class.java to Subtype.INTEGER,
        Byte::class.javaPrimitiveType!! to Subtype.INTEGER,
        Short::class.java to Subtype.INTEGER,
        Short::class.javaPrimitiveType!! to Subtype.INTEGER,
        Char::class.java to Subtype.CHAR,
        Char::class.javaPrimitiveType!! to Subtype.CHAR
)

/**
 * Creates a instance domainOf `SubType` based on a Class.
 *
 * @param typeClass
 * The class
 * @return The correct instance relative to the class.
 */
fun fromType(typeClass: Class<*>): Subtype {
    if (typeClass.isEnum) {
        return Subtype.STRING
    }
    val simpleType = INTERNAL_MAP[typeClass]
    return simpleType ?: Subtype.OBJ
}

fun isSimple(type: Class<*>): Boolean {
    return INTERNAL_MAP.containsKey(type) || type.isEnum
}

